plugins/plugin-optimize/plugin.js

```diff
@@ -0,0 +1,192 @@
+const fs = require('fs');
+const path = require('path');
+const glob = require('glob');
+const colors = require('kleur/colors');
+const esbuild = require('esbuild');
+const {init, parse} = require('es-module-lexer');
+const PQueue = require('p-queue').default;
+const {
+  appendHTMLToBody,
+  appendHTMLToHead,
+  HTML_JS_REGEX,
+  isRemoteModule,
+  relativeURL,
+  removeLeadingSlash,
+} = require('./util');
+
+/**
+ * Default optimizer for Snawpack, unless another one is given
+ */
+exports.default = function plugin(config, userDefinedOptions) {
+  const options = {
+    minifyJS: true,
+    preloadModules: true,
+    ...(userDefinedOptions || {}),
+  };
+
+  const CONCURRENT_WORKERS = require('os').cpus().length;
+
+  /** Scan a JS file for static imports */
+  function scanForStaticImports({file, rootDir, scannedFiles, importList}) {
+    try {
+      // 1. scan file for static imports
+      scannedFiles.add(file); // mark file as scanned
+      importList.add(file); // also mark file as an import if it hasn’t been already
+      let code = fs.readFileSync(file, 'utf-8');
+      const [imports] = parse(code);
+      imports
+        .filter(({d}) => d === -1) // this is where we discard dynamic imports (> -1) and import.meta (-2)
+        .forEach(({s, e}) => {
+          const specifier = code.substring(s, e);
+          importList.add(
+            specifier.startsWith('/')
+              ? path.join(rootDir, removeLeadingSlash(file))
+              : path.resolve(path.dirname(file), specifier),
+          );
+        });
+
+      // 2. recursively scan imports not yet scanned
+      [...importList]
+        .filter((fileLoc) => !scannedFiles.has(fileLoc)) // prevent infinite loop
+        .forEach((fileLoc) => {
+          scanForStaticImports({file: fileLoc, rootDir, scannedFiles, importList}).forEach(
+            (newImport) => {
+              importList.add(newImport);
+            },
+          );
+        });
+
+      return importList;
+    } catch (err) {
+      console.warn(
+        colors.dim('[@snowpack/plugin-optimize]') +
+          colors.yellow(
+            ` module preload failed: could not locate "${path.relative(rootDir, file)}"`,
+          ),
+      );
+      return importList;
+    }
+  }
+
+  /** Given a set of HTML files, trace the imported JS */
+  function preloadModulesInHTML(htmlFile, rootDir) {
```

Comment:
one thing that came up in another issue/discussion: what if we generated this mapping inside of Snowpack, ahead of time, and then exposed it to optimize plugins? 

I guess it would only really be guaranteed for the first optimize() plugin, since that could make changes that then wouldn't be reflected for the second one.

Just a thought, not something to block or even really address now if we wanted to.